workflows/pipe-common/scripts/autoscale_sge.py

```diff
@@ -1300,10 +1317,115 @@ def _validate_existence(self, host, hosts_dict):
             raise ScalingError('Host with name \'%s\' doesn\'t exist in the host storage' % host)
 
 
+class LastActionMarker:
+
+    def __init__(self):
+        self.last_action_timestamp = None
+        self.last_tag_timestamp = None
+
+
+class GridEngineWorkerTagsHandler:
+    _WORKER_TAG = 'SGE_IN_USE'
+
+    def __init__(self, api, tagging_active_timeout, host_storage, static_host_storage, clock, common_utils):
+        """
+        Processes active additional workers tags: if at least one job is running at the additional host
+        the corresponding run shall be tagged.
+
+        :param api: Cloud pipeline client.
+        :param tagging_active_timeout: Indicates how many seconds must pass before the run is recognized as active.
+        :param host_storage: Additional hosts storage.
+        :param static_host_storage: Static workers host storage.
+        :param clock: Clock.
+        :param common_utils: helpful stuff.
+        """
+        self.api = api
+        self.host_storage = host_storage
+        self.static_host_storage = static_host_storage
+        self.clock = clock
+        self.last_monitored_hosts = {}
+        self.tagging_active_timeout = timedelta(seconds=tagging_active_timeout)
+        self.common_utils = common_utils
+        self.static_hosts = self.static_host_storage.load_hosts()
+
+    def process_tags(self):
+        try:
+            Logger.info('Start tags processing step at %s.' % self.clock.now())
+            current_hosts = self.host_storage.load_hosts()
+            hosts_activity = self.host_storage.get_hosts_activity(current_hosts)
+            hosts_activity.update(self.static_host_storage.get_hosts_activity(self.static_hosts))
+            monitored_hosts = list(self.last_monitored_hosts.keys())
+            current_hosts += self.static_hosts
+            self._process_current_hosts(current_hosts, hosts_activity)
+            self._process_outdated_hosts(monitored_hosts, current_hosts)
+            Logger.info('Finish tags processing step at %s.' % self.clock.now())
+        except Exception as e:
+            Logger.warn('Tags processing has failed: %s' % str(e))
```

Comment:
Let's log error stacktraces as well.

```suggestion
        except Exception as e:
            Logger.warn('Tags processing has failed: %s' % str(e))
            Logger.warn(traceback.format_exc())
```